cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/InMemoryProgramRuntimeService.java

```diff
@@ -77,9 +77,31 @@ public synchronized RuntimeInfo run(Program program, ProgramOptions options) {
 
       Program bundleJarProgram = Programs.createWithUnpack(program.getJarLocation(), destinationUnpackedJarDir);
       RuntimeInfo info = super.run(bundleJarProgram, options);
-      info.getController().addListener(new AbstractListener() {
+      final ProgramController controller = info.getController();
+      controller.addListener(new AbstractListener() {
+
+        @Override
+        public void init(ProgramController.State state) {
+          if (state == ProgramController.State.COMPLETED) {
```

Comment:
I've seen it couple times already. Maybe make it ask part of `AbstractListener()` so that sub-classes don't need to always override it. Maybe sure find all sub-classes of `AbstractListener` when refactoring this.